packages/upload-api/src/blob/replicate.js

```diff
@@ -0,0 +1,188 @@
+import * as Server from '@ucanto/server'
+import { Delegation, Message } from '@ucanto/core'
+import * as Validator from '@ucanto/validator'
+import * as Transport from '@ucanto/transport/car'
+import * as SpaceBlob from '@storacha/capabilities/space/blob'
+import * as BlobReplica from '@storacha/capabilities/blob/replica'
+import * as Assert from '@web3-storage/content-claims/capability/assert'
+import * as Digest from 'multiformats/hashes/digest'
+import { base58btc } from 'multiformats/bases/base58'
+import { equals } from 'multiformats/bytes'
+import * as DID from '@ipld/dag-ucan/did'
+import * as API from '../types.js'
+import { AgentMessage } from '../lib.js'
+import { isBlobReplicaTransfer, toLocationCommitment } from './lib.js'
+import { createConcludeInvocation } from '../ucan/conclude.js'
+
+/**
+ * @param {API.BlobServiceContext} context
+ * @returns {API.ServiceMethod<API.SpaceBlobReplicate, API.SpaceBlobReplicateSuccess, API.SpaceBlobReplicateFailure>}
+ */
+export const blobReplicateProvider = (context) => {
+  const { router, registry, replicaStore, agentStore, maxReplicas } = context
+  return Server.provideAdvanced({
+    capability: SpaceBlob.replicate,
+    handler: async ({ capability, invocation, context: invContext }) => {
+      const { with: space, nb } = capability
+
+      if (nb.replicas > maxReplicas) {
+        return Server.error(/** @type {API.ReplicationCountRangeError} */ ({
+          name: 'ReplicationCountRangeError',
+          message: `requested number of replicas is greater than maximum: ${maxReplicas}`
+        }))
+      }
+
+      const digest = Digest.decode(nb.blob.digest)
+      const findRes = await registry.find(space, digest)
+      if (findRes.error) {
+        if (findRes.error.name === 'EntryNotFound') {
+          return Server.error(/** @type {API.ReplicationSourceNotFound} */ ({
+            name: 'ReplicationSourceNotFound',
+            message: `blob not found: ${base58btc.encode(digest.bytes)} in space: ${space}`
+          }))
+        }
+        return findRes
+      }
+
+      // check if we have any existing replicas
+      const replicaListRes = await replicaStore.list({ space, digest })
+      if (replicaListRes.error) {
+        return replicaListRes
+      }
+
+      // TODO: handle the case where a receipt was not received and the replica
+      // still exists in "allocated", but has actually timed out/failed.
+      const existingReplicas = replicaListRes.ok.filter(r => r.status !== 'failed')
+      if (existingReplicas.length + nb.replicas > maxReplicas) {
+        return Server.error(/** @type {API.ReplicationCountRangeError} */ ({
+          name: 'ReplicationCountRangeError',
+          message: `total number of replicas is greater than maximum: ${maxReplicas}`
```

Comment:
for this case it might be nice to include the number of existing replicas in the error.